### PR TITLE
Create representable capabilities in the image activation code

### PR DESCRIPTION
**Pull request by arichardson**
Currently the image activation code doesn't add padding/re-align the load address
when loading large executables. This can result in address space that is covered by 
AT_BASE, etc. being usuble by mmap() and can also result in incorrect capabilities
being passed to rtld that do not cover the mapped area (since the prog_cap/interp_cap
helper functions blindly align the base and length).

The actual fixes are the last three commits, but I ended up cleaning up various things
while testing this code. This patch series also removes the old hack of passing
relocbase in $c4 for purecap MIPS and instead loads it from auxargs like RISC-V.

### CODE REVIEW

File: sys/kern/imgact_elf.c [hunk 1]
```diff
@@ -1000,34 +1083,28 @@ __CONCAT(rnd_, __elfN(base))(vm_map_t map __unused, u_long minv, u_long maxv,
 	return (res);
 }
 
-static int
-__elfN(enforce_limits)(struct image_params *imgp, const Elf_Ehdr *hdr,
+static __attribute((optnone)) int
```
Comment (brooksdavis) [inline sys/kern/imgact_elf.c:1086]: debugging leftover?

File: sys/kern/imgact_elf.c [hunk 2]
```diff
@@ -629,9 +631,9 @@ __elfN(map_insert)(struct image_params *imgp, vm_map_t map, vm_object_t object,
 	return (KERN_SUCCESS);
 }
 
-static int
-__elfN(load_section)(struct image_params *imgp, vm_ooffset_t offset,
-    caddr_t vmaddr, size_t memsz, size_t filsz, vm_prot_t prot)
+static int __elfN(load_section)(const struct image_params *imgp,
```
Comment (brooksdavis) [inline sys/kern/imgact_elf.c:634]:
```suggestion
static int
__elfN(load_section)(const struct image_params *imgp,
```

File: sys/kern/imgact_elf.c [hunk 3]
```diff
@@ -2860,9 +3046,8 @@ __elfN(note_procstat_auxv)(void *arg, struct sbuf *sb, size_t *sizep)
 	}
 }
 
-static boolean_t
-__elfN(parse_notes)(struct image_params *imgp, Elf_Note *checknote,
-    const char *note_vendor, const Elf_Phdr *pnote,
+static boolean_t __elfN(parse_notes)(const struct image_params *imgp,
```
Comment (brooksdavis) [inline sys/kern/imgact_elf.c:3049]:
```suggestion
static boolean_t
__elfN(parse_notes)(const struct image_params *imgp,
```

**Review by brooksdavis — COMMENTED** [review]
I'm almost certain this will conflict with pure capability kernel changes. I've added @qwattash to the reviewers list.

File: sys/kern/imgact_elf.c (hunk 1 above)
Comment (arichardson) [reply sys/kern/imgact_elf.c:1086]: Yeah clang was optimizing out too many useful variables...

File: libexec/rtld-elf/rtld.c [hunk 4]
```diff
@@ -1883,7 +1883,12 @@ digest_phdr(const Elf_Phdr *phdr, int phnum, dlfunc_t entry, const char *path)
     }
 
 #ifdef __CHERI_PURE_CAPABILITY__
-   obj->relocbase = cheri_setbounds(obj->relocbase, obj->mapsize);
+    if (cheri_bytes_remaining(obj->relocbase) > CHERI_REPRESENTABLE_LENGTH(
+	obj->mapsize)) {
```
Comment (arichardson) [inline libexec/rtld-elf/rtld.c:1887]: rtld.c uses really messed up indentation, but I think this is consistent with the rest of the file...

**Comment by arichardson** [comment]
I just had a look at the diff to the purecap kernel: Most of this should still apply, but the reservation change will no longer be needed.

The new function __elfN(ensure_representable_base) can be merged with __elfN(build_imgact_capability).

If we think that the reservations changes will land soon, I can drop the "reserve space commit" to reduce churn.

**Comment by nwf** [comment]
This works for me in local testing, thanks!  It warns `Attempted to load /mnt/snmalloc-mips64-purecap-build/func-malloc-1 with non-representable mapping: moving from 0x120000 to 0x180000` which seems plausible.

File: sys/kern/imgact_elf.c [hunk 5]
```diff
@@ -534,9 +536,9 @@ __elfN(map_partial)(vm_map_t map, vm_object_t object, vm_ooffset_t offset,
 	/*
 	 * Create the page if it doesn't exist yet. Ignore errors.
 	 */
-	vm_map_fixed(map, NULL, 0, trunc_page(start), round_page(end) -
-	    trunc_page(start), VM_PROT_ALL, VM_PROT_ALL,
-	    MAP_CHECK_EXCL);
+	vm_map_fixed(map, NULL, 0, trunc_page(start),
+	    round_page(end) - trunc_page(start), prot,
+	    prot /* XXX: or VM_PROT_ALL? */, MAP_CHECK_EXCL);
```
Comment (qwattash) [inline sys/kern/imgact_elf.c:541]: I think maxprot should be fine as not being set to `VM_PROT_ALL`

**Comment by qwattash** [comment]
> I just had a look at the diff to the purecap kernel: Most of this should still apply, but the reservation change will no longer be needed.
> 
> The new function __elfN(ensure_representable_base) can be merged with __elfN(build_imgact_capability).
> 
> If we think that the reservations changes will land soon, I can drop the "reserve space commit" to reduce churn.

The `build_imgact_capability` function should take care of the whole reservation issue, but the approach is slightly different. We first reserve the whole space needed and then map the various sections using the reservation capability that is created by `build_imact_capability`.
I don't have a specific ETA for landing the reservation patch, I think I can work on it next, it should only have dependency on the `vm_map` changes and cheric macros.

**Comment by arichardson** [comment]
@qwattash @brooksdavis should I drop the change to reserve address space to avoid conflicts? Or we could revert it before merging the purecap imgact reservation changes?
It seems unlikely that someone could exploit it but avoiding mappings inside another DSO's PCC is a security fix and should probably be merged before the next release?

**Comment by brooksdavis** [comment]
Reverting the reservation as part of the merge seems likely to be ok as long as purecap reservations are going to land soon.

**Pull request merged**